### PR TITLE
Json

### PR DESCRIPTION
**Pull request by stepan-tsirkin**
This PR is work in progress, to implement the output to json format. 

 I am opening it to inspire discussion

### CODE REVIEW

**Comment by stepan-tsirkin** [comment]
So far I just implemented storing spacegroup. 

{"hello": "world", "spacegroup": {"name": "R-3m", "number": 166, "num_symmetries": 12, "symmetries": {"1": {"calculation cell coincides with reference cell": false, "rotation_matrix": {"@module": "numpy", "@class": "array", "dtype": "int32", "data": [[1, 0, 0], [0, 1, 0], [0, 0, 1]]}, "rotation_matrix_refUC": {"@module": "numpy", "@class": "array", "dtype": "int64", "data": [[1, 0, 0], [0, 1, 0], [0, 0, 1]]}, "translation": {"@module": "numpy", "@class": "array", "dtype": "float64", "data": [0.0, 0.0, 0.0]}, "translation_refUC": {"@module": "numpy", "@class": "array", "dtype": "float64", "data": [0.0, 0.0, 0.0]}, "axis": {"@module": "numpy", "@class": "array", "dtype": "float64", "data": [0.028867072380382164, -0.8113818220445695, -0.5838032468116466]}, "angle": " 0 ", "inversion": false, "sign": 1}, "2": {"calculation cell coincides with reference cell": false, "rotation_matrix": {"@module": "numpy", "@class": "array", "dtype": "int32", "data": [[0, 0, 1], [1, 0, 0], [0, 1, 0]]}, "

The idea is to store all needed data into one nested dictionary, and finally store it with  `monty.serialization.dumpfn` . I have some questions:

1.  The arrays are stored as `{"@module": "numpy", "@class": "array", "dtype": "int32", "data": [[1, 0, 0], [0, 1, 0], [0, 0, 1]]}`  Is it good for compatibility with other languages? Or is it better to convertt arrays into nested lists? Is there a laconic way to do it? 
2.  can dumpfn  write it in a more human-readable way, with more linebreaks, tabulation, etc., so that the structure of the dictiornaries is more visible by eye? 

@mkhorton @miraola

**Comment by mkhorton** [comment]
1. The arrays are stored as {"@module": "numpy", "@class": "array", "dtype": "int32", "data": [[1, 0, 0], [0, 1, 0], [0, 0, 1]]} Is it good for compatibility with other languages? Or is it better to convertt arrays into nested lists? Is there a laconic way to do it?

Yes, you can call `.tolist()` on the array if preferred; this would store it in JSON as a list of floats. However, the advantage of the current method is that this additionally specifies the datatype (`int32`) of the numpy array, so less information is lost.

2. Yes, you can add `indent=4` to `dumpfn` to add indentations.

In both cases, the intent of the JSON output is for machine readability however, with the code deciding how to turn that into a human-readable string, so I don't think it's a big problem if it's not readable.

**Comment by stepan-tsirkin** [comment]
> Yes, you can call .tolist() on the array if preferred; this would store it in JSON as a list of floats. 

Yes, but that's smth to be done on every addition of an array to the dictionary., I thought there might be an option in dumpfn that does it automatically for all arrays

>However, the advantage of the current method is that this additionally specifies the datatype (int32) of the numpy array, so less information is lost.

> I see/ actually, if one wants to get only the data as list, they are in the data field. So, it is anyway more convenient to store as arrays.

> In both cases, the intent of the JSON output is for machine readability however, 

Yes, just for someone who is lazy to read documentation (which we are also lazy to write) looking at the fields of the json file already gives an impression on what data is stored there.

**Comment by stepan-tsirkin** [comment]
Hi @mkhorton , @MIraola 

I implemented writing to json, do you have any comments, did I miss any essential info?

Maybe the next step would be to modify the tests sos that ttey use the info from json instead of parsing output. 

@mkhorton , can you do this? You may take the old tests, and modify them, so that essentially they check the ame info, but form json. And we may keep both sets of tests so far.

**Comment by mkhorton** [comment]
> Maybe the next step would be to modify the tests sos that ttey use the info from json instead of parsing output.

> @mkhorton , can you do this? You may take the old tests, and modify them, so that essentially they check the ame info, but form json. And we may keep both sets of tests so far.

Yes, this sounds sensible and I'd be happy to. I can't give a time commitment right now however.

**Comment by stepan-tsirkin** [comment]
>Yes, this sounds sensible and I'd be happy to. I can't give a time commitment right now however.

Ok, I think we can resubmit without it. @Mikel, if you are fine with the changes, just merge.

File: irrep/bandstructure.py [hunk 1]
```diff
@@ -1314,6 +1329,11 @@ def write_trace_all(
         """
         f = open(fname, "w")
         kpline = self.KPOINTSline()
+        json_data = {}
```
Comment (MIraola) [inline irrep/bandstructure.py:1332]: Is this json object used somewhere? It is not returned by the function, neither saved before the function reaches the end.

Comment (stepan-tsirkin) [reply irrep/bandstructure.py:1332]: Yes, this one should be removed

**Pull request merged**